Articles/Blog/TeamExperiencesImplementingCI.md

```diff
@@ -0,0 +1,82 @@
+
+[Mark C. Miller](https://github.com/markcmiller86) observed that “costs and time limits were some of the key factors affecting our CI provider choice.”  One of the largest continuing challenges faced by the team is the compute time required to build and run the tests.  “Dependencies aside, it takes a 2017 4-core MacBook-Pro about 15 mins to build VisIt using `make -j 8` (oversubscription by 2x). It takes another 90+ mins to build dependencies. The entire nightly test suite itself can take 2-4 hours.” Even though Azure Pipelines provides unlimited run times for open-source projects, there is still a practical limit on what can be effectively tested in CI.
+
+This has led VisIt’s CI to utilize mainly the command-line controller.  As a Python interface to a single executable, tests can exercise different code paths for things like plots, database readers, operators, rendering modalities, etc., without requiring a complete rebuild of VisIt.  The team also uses stand-alone tests of some key classes.  Substantial unit testing, however, would require refactoring the core of VisIt to support that modality of build and execution.
+
+Miller notes there’s always a trade-off between better testing versus improving the product itself.  There’s a laundry list of functionality for which the project would like to have more testing, like GUI events, multiple client/server modalities, numerical corner cases like NaN/Inf floating-point exception handling, and more comprehensive doc-tests.  Rather than trying to tackle all of these directly, the developers have found that regular discussions comparing potential improvements have made them extremely productive.  Discussing ideas at length allows the team to document ideas via GitHub issues, understand tradeoffs, and prioritize their work.
+
+Miller also had a lot of advice for teams developing an initial process for CI.  As a guiding principle, he advocates for starting small and aiming for the ultimate goal of moving a subset of tests into CI — ones that are responsive, informative, and cover areas that the team agree should be continuously checked.  Those tests should use the code in the way its intended for users to run, minimizing surprises down the road.
+
+VisIt’s development path was heavily influenced by the prior experience of its developers with CI, along with a healthy dose of [combing through documentation](https://bssw.io/blog_posts/continuous-technology-refreshment-an-introduction-using-recent-tech-refresh-experiences-on-visit).  Perhaps uniquely, the team uses CI primarily as a “smoke test” for compilation, dependencies, and their release process, then place a majority of their software quality focus on reviewing pull requests and running nightly tests on local resources.  Those nightly tests generate, then check 3500+ images and 2000+ numbers/textual results.  Although the tests require bit-for-bit accuracy, the team has a rolling error-resolution process.
+
+### Case 3: ZFP
```

Comment:
@frobnitzem in the other two cases, you explicitly name 1-2 people, but no one here.  For symmetry, is there someone you could name here too?

Reply:
I added Peter Lindstrom (whom Mark interviewed) to the ack-s.